### ISSUE TITLE
Cannot create log appender - exception thrown from ctor

### ISSUE THREAD

**Issue by plmwong** [plmwong/log4net.Raygun#3]
The following exception is thrown when attempting to log an error:

```
#!c#
log4net:ERROR XmlHierarchyConfigurator: Could not create Appender [RaygunAppender] of type [log4net.Raygun.RaygunAppender, log4net.Raygun]. Reported error follows.
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Configuration.ConfigurationErrorsException: The configuration is read only.
   at System.Configuration.ConfigurationElement.SetPropertyValue(ConfigurationProperty prop, Object value, Boolean ignoreLocks)
   at Mindscape.Raygun4Net.RaygunSettings.set_ThrowOnError(Boolean value) in e:\Users\Jason\Documents\GitHub\raygun4net\Mindscape.Raygun4Net\RaygunSettings.cs:line 43
   at log4net.Raygun.RaygunAppender..ctor() in c:\Users\pwong\Dev\log4net.Raygun\log4net.Raygun\RaygunAppender.cs:line 23
   --- End of inner exception stack trace ---
   at System.RuntimeTypeHandle.CreateInstance(RuntimeType type, Boolean publicOnly, Boolean noCheck, Boolean& canBeCached, RuntimeMethodHandleInternal& ctor, Boolean& bNeedSecurityCheck)
   at System.RuntimeType.CreateInstanceSlow(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, StackCrawlMark& stackMark)
   at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean skipCheckThis, Boolean fillCache, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, Boolean nonPublic)
   at System.Activator.CreateInstance(Type type)
   at log4net.Repository.Hierarchy.XmlHierarchyConfigurator.ParseAppender(XmlElement appenderElement)

```

I have already opened an issue at Raygun repo: https://github.com/MindscapeHQ/raygun4net/issues/184

----------------------------------------
- Bitbucket: https://bitbucket.org/plmw/log4net.raygun/issue/3
- Originally reported by: Anonymous
- Originally created at: 2014-10-08T16:23:28.905

**Comment by plmwong** [comment]
Hiya,

As you've probably already surmised, this happens because the `RaygunSettings` collection has already been created and the `ThrowOnError` property fails to be set.

Do you have the `RaygunSettings` configuration section declared at the top of your configuration file? E.g.

```
<section name="RaygunSettings" type="Mindscape.Raygun4Net.RaygunSettings, Mindscape.Raygun4Net"/>
```

Unless you have a real need to use that configuration section (settings such as `apiKey` and the `ignoredFormNames` etc. should be set on the log4net appender configuration instead), then you can remove the RaygunSettings section declaration entirely, and this error should go away.

If you do need one of the other raygun settings (e.g. `mediumTrust`, `endPoint`), then please let me know.


Cheers,

Phil

----------------------------------------
Original comment by: Phillip Wong

**Comment by plmwong** [comment]
Hi!

Removing the `RaygunSettings` section indeed prevents the exception from being thrown and being able to use your log appender. However, it also prevents RaygunErrorModule from catching/sending unhandled exceptions to raygun.io.

I see that the log appender's retry logic needs an exception to be thrown if Raygun fails to send the error. Since the `retries` feature is optional, enabling Raygun's `ThrowOnError` should also only happen if the user wants to use that feature? 
As an alternative, the `retries` feature could require the `ThrowOnError` being enabled in `RaygunSettings` section which in my opinion is better than enabling (or at least trying to) it silently.

In any case, if Raygun accepts my suggestion, it will be possible to both have the `RaygunSettings` section and set this property in the log appender. 

Thanks,
Girts

----------------------------------------
Original comment by: [girtsl](http://bitbucket.org/girtsl)

**Comment by plmwong** [comment]
I think I would prefer to not require a RaygunSettings configuration section declaration just to use retries. I'll probably look into making a couple of changes along the lines of:

* Only look for `ThrowOnError` if `retries` is enabled;
* If `retries` is enabled, then if the `RaygunSettings` are ReadOnly (i.e. when already initialised from the configuration section), look at whether `ThrowOnError` is set to true and if not, throw a custom exception informing the user that it is required to be enabled for retries.

Thanks,
Phil

----------------------------------------
Original comment by: Phillip Wong

**Comment by plmwong** [comment]
Your proposed changes are very welcome and it looks that they will resolve the issue. 

Thanks,
Girts

----------------------------------------
Original comment by: [girtsl](http://bitbucket.org/girtsl)

**Comment by plmwong** [comment]
Hi,

Do you plan to merge the new `fix/throwonerror` branch with master and release it any time soon? :)

Cheers,
Girts

----------------------------------------
Original comment by: [girtsl](http://bitbucket.org/girtsl)

**Comment by plmwong** [comment]
Hey,

Sorry, been out of the country the past week... hopefully will get this done in the next day or two :)

Cheers,

Phil

----------------------------------------
Original comment by: Phillip Wong

**Comment by plmwong** [comment]
Should be resolved now in v2.4. Just pushed up to nuget.

----------------------------------------
Original comment by: Phillip Wong

**Comment by plmwong** [comment]
Hi,
Works great!

Thanks,
Girts

----------------------------------------
Original comment by: [girtsl](http://bitbucket.org/girtsl)
